### ISSUE TITLE
Systemctl start service in Redhat with SElinux set to Enforcing fails

### ISSUE THREAD

**Issue by Veverke** [dotnet/runtime#56729]
*Edit by @antonfirsov : after resolving the dotnet-dump issue, new stack-trace here: https://github.com/dotnet/runtime/issues/56729#issuecomment-891641187*

----

**I have a dotnet core (3.1) worker service implemented as a systemd daemon in Redhat, SElinux enabled and set to Enforcing.**
Running **`systemctl start [myService]`** sometimes fails with `systemd[1]: [myService]: Failed with result 'protocol' ` (taken from the output of `journalctl -xe -u [myService]`)
**`systemctl status [myService]`** shows  `Active: activating (auto-restart) (Result: protocol)` (it auto restarts every 30 seconds)

I managed to run a dotnet-dump on the executable process id this systemd executes (defined by `ExecStart`). The output is as follows:

``` 
Writing full to /home/redhat-build/Downloads/core_20210802_174944
Unhandled exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (13): **Permission denied /tmp/dotnet-diagnostic-22279-705688-socket**
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 4296
   at System.Net.Sockets.Socket.Connect(EndPoint remoteEP) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 892
   at System.Net.Sockets.Socket.UnsafeBeginConnect(EndPoint remoteEP, AsyncCallback callback, Object state, Boolean flowContext) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 2151
   at System.Net.Sockets.Socket.BeginConnect(EndPoint remoteEP, AsyncCallback callback, Object state) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 2122
   at Microsoft.Diagnostics.NETCore.Client.IpcSocket.Connect(EndPoint remoteEP, TimeSpan timeout) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcSocket.cs:line 50
   at Microsoft.Diagnostics.NETCore.Client.IpcUnixDomainSocket.Connect(EndPoint localEP, TimeSpan timeout) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcUnixDomainSocket.cs:line 31
   at Microsoft.Diagnostics.NETCore.Client.IpcEndpointHelper.Connect(IpcEndpointConfig config, TimeSpan timeout) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:line 69
   at Microsoft.Diagnostics.NETCore.Client.PidIpcEndpoint.Connect(TimeSpan timeout) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:line 230
   at Microsoft.Diagnostics.NETCore.Client.IpcClient.SendMessage(IpcEndpoint endpoint, IpcMessage message) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs:line 25
   at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.WriteDump(DumpType dumpType, String dumpPath, Boolean logDumpGeneration) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/DiagnosticsClient.cs:line 101
   at Microsoft.Diagnostics.Tools.Dump.Dumper.Collect(IConsole console, Int32 processId, String output, Boolean diag, DumpTypeOption type, String name) in /_/src/Tools/dotnet-dump/Dumper.cs:line 111
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) in /_/src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs:line 399
   at System.Delegate.DynamicInvokeImpl(Object[] args) in /_/src/coreclr/src/System.Private.CoreLib/src/System/Delegate.CoreCLR.cs:line 89
   at System.Delegate.DynamicInvoke(Object[] args) in /_/src/libraries/System.Private.CoreLib/src/System/Delegate.cs:line 64
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__21_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__22_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__20_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

**The dump shows an access denied at /tmp/dotnet-diagnostic...-socket**
I have seen SELinux issues with docker sockets before. Is this related ?
Does the solution rely in declaring a SElinux permission policy that will allow access to this file ? What are the files, however, that are being accessed in starting a systemd daemon ?

**Comment by antonfirsov** [comment]
@Veverke the stack trace above is coming from `dotnet-dump`, so I'm afraid we don't see the cause of your original problem.

Can you please open an issue against https://github.com/dotnet/diagnostics first? They will hopefully help resolving the permission issues with `dotnet-dump`, so you can investigate the dump and raise a new issue here, if necessary.

I would transfer this issue, but for some reason GitHub doesn't let me pick dotnet/diagnostics as target repository :(

**Comment by wfurt** [comment]
What is your policy @Veverke ? With `Enforcing` the process may not be able to make system calls it needs. You will need to review your logs and policy and perhaps make adjustments. The default set may not be what .NET needs. 

cc: @tmds

**Comment by hoyosjs** [comment]
Not sure what will the dump be able to tell you, but the error in dotnet-dump happens because you are trying to dump a process you didn't start. The same user needs to trigger dotnet-dump, or you need to run it with sudo. The traced process can't have APIs like ptrace restricted either.

**Comment by Veverke** [comment]
@hoyosjs: [this article](https://devblogs.microsoft.com/dotnet/collecting-and-analyzing-memory-dumps/) seem to state dotnet-dump does not require being executed with sudo anymore. Nevertheless, I tried what you did and am now working on how to read the generated dump file (looks like you are right in that the output I got, originally, is human readable, because it comes from an exception **during** the dump file generation...)

**Comment by Veverke** [comment]
Following up on @antonfirsov and @hoyosjs remarks, I tried once again to extract a dotnet-dump while my systemd daemon crashes (as instructed [here](https://github.com/dotnet/diagnostics/blob/main/documentation/dotnet-dump-instructions.md)), and the results are pasted below. Most of the data is not new - this is more or less the dotnet exception trace that journalctl -xe -u [myService] displays - it adds a few lower level info.

OS Thread Id: 0x10a6a (0)
        Child SP               IP Call Site
00007FFD7B5E9C58 00007f2c6a532805 [InlinedCallFrame: 00007ffd7b5e9c58] Interop+Sys.OpenDir(System.String)
00007FFD7B5E9C58 00007f2bf06657da [InlinedCallFrame: 00007ffd7b5e9c58] Interop+Sys.OpenDir(System.String)
00007FFD7B5E9C40 00007F2BF06657DA ILStubClass.IL_STUB_PInvoke(System.String)
00007FFD7B5E9D00 00007F2BF08A7E71 System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib]].CreateDirectoryHandle(System.String, Boolean) [/_/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs @ 72]
00007FFD7B5E9D40 00007F2BF08A7BE1 System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib]].Init() [/_/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs @ 40]
00007FFD7B5E9D80 00007F2BF08A7988 System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib]]..ctor(System.String, Boolean, System.IO.EnumerationOptions) [/_/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.cs @ 42]
00007FFD7B5E9DB0 00007F2BF08A74AA System.IO.Enumeration.FileSystemEnumerable`1[[System.__Canon, System.Private.CoreLib]]..ctor(System.String, FindTransform<System.__Canon>, System.IO.EnumerationOptions, Boolean) [/_/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerable.cs @ 39]
00007FFD7B5E9DF0 00007F2BF089927B System.IO.Enumeration.FileSystemEnumerableFactory.UserDirectories(System.String, System.String, System.IO.EnumerationOptions) [/_/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerableFactory.cs @ 142]
00007FFD7B5E9E20 00007F2BF089C6FB System.IO.Directory.InternalEnumeratePaths(System.String, System.String, System.IO.SearchTarget, System.IO.EnumerationOptions) [/_/src/System.IO.FileSystem/src/System/IO/Directory.cs @ 187]
00007FFD7B5E9E50 00007F2BF089F48A System.IO.Directory.EnumerateDirectories(System.String) [/_/src/System.IO.FileSystem/src/System/IO/Directory.cs @ 195]
00007FFD7B5E9E60 00007F2BF325EC55 System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 445]
00007FFD7B5E9FD0 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EA140 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EA2B0 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EA420 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EA590 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EA700 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EA870 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EA9E0 00007F2BF325ECBD System.IO.FileSystemWatcher+RunningInstance.AddDirectoryWatchUnlocked(WatchedDirectory, System.String) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 447]
00007FFD7B5EAB50 00007F2BF325E600 System.IO.FileSystemWatcher+RunningInstance..ctor(System.IO.FileSystemWatcher, Microsoft.Win32.SafeHandles.SafeFileHandle, System.String, Boolean, System.IO.NotifyFilters, System.Threading.CancellationToken) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 305]
00007FFD7B5EABA0 00007F2BF325DBFE System.IO.FileSystemWatcher.StartRaisingEvents() [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs @ 70]
00007FFD7B5EAC20 00007F2BF325D923 System.IO.FileSystemWatcher.StartRaisingEventsIfNotDisposed() [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs @ 645]
00007FFD7B5EAC40 00007F2BF325C3D6 System.IO.FileSystemWatcher.set_EnableRaisingEvents(Boolean) [/_/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs @ 155]
00007FFD7B5EAC70 00007F2BF0FA9C30 Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.TryEnableFileSystemWatcher() [/_/src/FileProviders/Physical/src/PhysicalFilesWatcher.cs @ 377]
00007FFD7B5EACA0 00007F2BF0FA8E31 Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.CreateFileChangeToken(System.String) [/_/src/FileProviders/Physical/src/PhysicalFilesWatcher.cs @ 135]
00007FFD7B5EACC0 00007F2BF0FA83F9 Microsoft.Extensions.FileProviders.PhysicalFileProvider.Watch(System.String) [/_/src/FileProviders/Physical/src/PhysicalFileProvider.cs @ 328]
00007FFD7B5EACE0 00007F2BF0D1430B Microsoft.Extensions.Configuration.FileConfigurationProvider.<.ctor>b__1_0() [/_/src/Configuration/Config.FileExtensions/src/FileConfigurationProvider.cs @ 36]
00007FFD7B5EACF0 00007F2BF0FEC448 Microsoft.Extensions.Primitives.ChangeToken+ChangeTokenRegistration`1[[System.__Canon, System.Private.CoreLib]]..ctor(System.Func`1<Microsoft.Extensions.Primitives.IChangeToken>, System.Action`1<System.__Canon>, System.__Canon) [/_/src/Primitives/src/ChangeToken.cs @ 71]
00007FFD7B5EAD20 00007F2BF0FE8951 Microsoft.Extensions.Primitives.ChangeToken.OnChange(System.Func`1<Microsoft.Extensions.Primitives.IChangeToken>, System.Action) [/_/src/Primitives/src/ChangeToken.cs @ 32]
00007FFD7B5EAD50 00007F2BF0D13DBB Microsoft.Extensions.Configuration.FileConfigurationProvider..ctor(Microsoft.Extensions.Configuration.FileConfigurationSource) [/_/src/Configuration/Config.FileExtensions/src/FileConfigurationProvider.cs @ 35]
00007FFD7B5EAD80 00007F2BF0CD40CA Microsoft.Extensions.Configuration.Json.JsonConfigurationSource.Build(Microsoft.Extensions.Configuration.IConfigurationBuilder) [/_/src/Configuration/Config.Json/src/JsonConfigurationSource.cs @ 21]
00007FFD7B5EADA0 00007F2BF0C858BC Microsoft.Extensions.Configuration.ConfigurationBuilder.Build() [/_/src/Configuration/Config/src/ConfigurationBuilder.cs @ 49]
00007FFD7B5EADD0 00007F2BF0DA9F6E Microsoft.Extensions.Hosting.HostBuilder.BuildAppConfiguration() [/_/src/Hosting/Hosting/src/HostBuilder.cs @ 198]
00007FFD7B5EAE10 00007F2BF0DA9AF9 Microsoft.Extensions.Hosting.HostBuilder.Build() [/_/src/Hosting/Hosting/src/HostBuilder.cs @ 130]
00007FFD7B5EAE30 00007F2BF0661A21 Sepio.Agent.Program.Main(System.String[]) [/home/redhat-build/Documents/source/repos/Sepio.Agent/Sepio.Agent/Program.cs @ 47]
00007FFD7B5EB198 00007f2c69a4d2df [GCFrame: 00007ffd7b5eb198] 
00007FFD7B5EB680 00007f2c69a4d2df [GCFrame: 00007ffd7b5eb680] 
> pe -lines

**Comment by Veverke** [comment]
> What is your policy @Veverke ? With `Enforcing` the process may not be able to make system calls it needs. You will need to review your logs and policy and perhaps make adjustments. The default set may not be what .NET needs.
> 
> cc: @tmds

Tomas, how do I query my current SElinux policies ? Or if that does not make sense, how do I answer your question ? Almost no knowledge on SElinux here.

**Comment by Veverke** [comment]
> @Veverke the stack trace above is coming from `dotnet-dump`, so I'm afraid we don't see the cause of your original problem.
> 
> Can you please open an issue against https://github.com/dotnet/diagnostics first? They will hopefully help resolving the permission issues with `dotnet-dump`, so you can investigate the dump and raise a new issue here, if necessary.
> 
> I would transfer this issue, but for some reason GitHub doesn't let me pick dotnet/diagnostics as target repository :(

@antonFirsov, should this still be moved to the [https://github.com/dotnet/diagnostics](https://github.com/dotnet/diagnostics) repo?

**Comment by antonfirsov** [comment]
@Veverke no, since we see a stack trace from your app now. Can you also share more detailed exception info (Exception type, message)? Any chance you can create and share a repro app?

Assigned to the `Extensions-Hosting` area for now.

**Comment by Veverke** [comment]
> Can you also share more detailed exception info (Exception type, message)? 
My dump analysis can't find any managed exceptions. So dotnet-dump can't help in providing more details, as far as I understand.

>Any chance you can create and share a repro app?
Can't see how. Will be a very cumbersome task.

**Comment by eerhardt** [comment]
Moving to `area-System.IO` since the exception is coming from `FileSystemWatcher`

**Comment by adamsitnik** [comment]
> sometimes fails with
> With Enforcing the process may not be able to make system calls it needs. 

In such a case, I would expect it to always fail with the same error.

> Interop+Sys.OpenDir(System.String)

@Veverke this part of your call stack is very interesting. Could you please use `strace` and find out what is being returned by `opendir` (this is the syscall that `OpenDir(System.String)` calls)

**Comment by adamsitnik** [comment]
> Microsoft.Extensions.Hosting

@eerhardt @davidfowl @Jozkee @carlossanlop  we should also most probably not fail the app startup if `FileConfigurationProvider` is unable to monitor for config file changes.

**Comment by eerhardt** [comment]
> we should also most probably not fail the app startup if FileConfigurationProvider is unable to monitor for config file changes.

I'm not sure. I think developers expect that behavior, and if we silently started skipping the file change monitoring when failed, they could become confused when they don't see their updates happening when they change the file. So I don't think this should be the default. I think the current behavior should be the default. And if we need to ignore the failures, that would be an opt-in mode.

Either way, that would be a separate issue.

**Comment by Veverke** [comment]
@adamsitnik : sorry for the delay, had context-switched to a different task. Context-switching back to this and... I do not know what to say. This issue aroused when trying to upgrade a .net app delivered as a daemon service (part of the job, besides upgrading the package itself, is to stop and start the service again, therefore me playing with systemctl stop/start commands). The issue is gone, I tried several times now (machine is probably after a boot, but I am sure I did this at the time). The only thing that may be different now (may be - again, I am pretty sure I tried this solution before opening this issue, but who daunts relying on his own memory these days...) - is that - after reading somewhere (pity I did not bookmark the post) that it is better not to deploy/install the app into /usr/lib, but rather to somewhere else, like /usr/local/bin. With my daemon spec file deploy dir set to /usr/local/bin I do not see this happening. (indeed, /usr/lib does not have (w) permissions to owner (root), while /usr/local/bin does).

I assume this will eventually be closed.
